src/python/grpcio_observability/grpc_observability/python_census_context.cc

```diff
@@ -0,0 +1,268 @@
+  memcpy(trace_option_rep_, &options, kTraceOptionsLen);
+
+  SpanContext context(absl::BytesToHexString(absl::string_view(reinterpret_cast<const char *>(trace_id_rep_), kTraceIdSize)),
+                      absl::BytesToHexString(absl::string_view(reinterpret_cast<const char *>(span_id_rep_), kSpanIdSize)),
+                      trace_option_rep_[0] & kIsSampled);
+  return context;
+}
+
+
+size_t TraceContextSerialize(PythonCensusContext& context,
+                             char* tracing_buf, size_t tracing_buf_size) {
+  if (tracing_buf_size < kGrpcTraceBinHeaderLen) {
+    return 0;
+  }
+  ToGrpcTraceBinHeader(context, reinterpret_cast<uint8_t*>(tracing_buf));
+  return kGrpcTraceBinHeaderLen;
+}
+
+
+size_t StatsContextSerialize(size_t /*max_tags_len*/, grpc_slice* /*tags*/) {
+  // TODO(unknown): Add implementation. Waiting on stats tagging to be added.
+  return 0;
+}
+
+
+size_t ServerStatsDeserialize(const char* buf, size_t buf_size,
+                              uint64_t* server_elapsed_time) {
+  return grpc::internal::RpcServerStatsEncoding::Decode(
+      absl::string_view(buf, buf_size), server_elapsed_time);
+}
+
+
+size_t ServerStatsSerialize(uint64_t server_elapsed_time, char* buf,
+                            size_t buf_size) {
+  return grpc::internal::RpcServerStatsEncoding::Encode(server_elapsed_time, buf,
+                                                  buf_size);
+}
+
+
+uint64_t GetIncomingDataSize(const grpc_call_final_info* final_info) {
+  return final_info->stats.transport_stream_stats.incoming.data_bytes;
+}
+
+
+uint64_t GetOutgoingDataSize(const grpc_call_final_info* final_info) {
+  return final_info->stats.transport_stream_stats.outgoing.data_bytes;
+}
+
+
+namespace {
+// span_id is a 16-character hexadecimal encoded string.
+std::string generateSpanId() {
+  uint64_t span_id = absl::Uniform<uint64_t>(absl::BitGen());
+  std::stringstream hex_string;
+  hex_string << std::setfill('0') << std::setw(16) << std::hex << span_id;
+  return std::string(hex_string.str());
+}
+
+// trace_id is a 32-character hexadecimal encoded string
+std::string generateTraceId() {
+  absl::uint128 trace_id = absl::Uniform<absl::uint128>(absl::BitGen());
+  std::stringstream hex_string;
+  hex_string << std::setfill('0') << std::setw(32) << std::hex << trace_id;
+  return std::string(hex_string.str());
+}
+
+} // namespace
+
+//
+// Span
+//
+
+Span Span::StartSpan(absl::string_view name, Span* parent) {
+  SpanContext context;
+  std::string span_id = generateSpanId();
+  std::string trace_id;
+  std::string parent_span_id;
+  bool should_sample;
+  auto start_time = absl::Now();
+
+  if (parent != nullptr) {
+    parent_span_id = parent->Context().SpanId();
+    trace_id = parent->Context().TraceId();
+    should_sample = parent->Context().IsSampled();
+  } else {
+    trace_id = generateTraceId();
+    should_sample = ShouldSample(trace_id);
+  }
+
+  context = SpanContext(trace_id, span_id, should_sample);
```

Comment:
Nit: You could just declare `context` here instead of default constructing then overwriting.